[SLP] SLP's copyable elements based upon Main/Alt operations.

dtemirbulatov opened the pull request:
Added testcase from https://github.com/llvm/llvm-project/issues/110740. It looks like there are no SLP schedular issues left, I was able to compiler and run LNT, Spec2k6,  Spec2k17 without any runtime/compile time issues with different flags "-O3, Ofast, -mllvm -slp-threshold=-99999".

alexey-bataev commented:
I foresee that the current scheduling model won't allow to fix all bugs with it. Need to fix it first to allow multi-scheduling of the instructions, otherwise, there always will be bugs

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -17114,17 +17529,64 @@ void BoUpSLP::optimizeGatherSequence() {
            "no ScheduleData for bundle member "
            "(maybe not in same basic block)");
-    assert(BundleMember->isSchedulingEntity() &&
+    assert(BundleMember->isSchedulingEntity() ||
```
This is wrong

dtemirbulatov replied:
Thanks for finding this.

dtemirbulatov replied:
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -4296,6 +4535,10 @@ class BoUpSLP {
     DenseMap<Instruction *, ScheduleData *> ScheduleDataMap;
 
+    /// Attaches ScheduleData to Instruction with the leading key.
+    DenseMap<Value *, SmallDenseMap<Value *, ScheduleData *>>
```
1. Not sure this solution will allow to solve all the problems.
2. Better to avoid adding new storages with the same logic as the existing one (ScheduleDataMap in your case). Better to extend the existing with the new functionality to avoid complex handling

dtemirbulatov replied:
Yes, thanks the single one storage is better.

dtemirbulatov replied:
Done.

dtemirbulatov replied:
Tested this solution, I could not find any more scheduling issues.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3298,6 +3462,7 @@ class BoUpSLP {
                          ///< complex node like select/cmp to minmax, mul/add to
                          ///< fma, etc. Must be used for the following nodes in
                          ///< the pattern, not the very first one.
+      CopyableVectorize, ///< The node for copyable elements.
```
Why do you need this new state?

dtemirbulatov replied:
My motivation was to simplifiy the solution, I will check the other way, thanks.

dtemirbulatov replied:
Done.

dtemirbulatov committed:
[SLP] SLP's copyable elements based upon Main/Alt operations.

dtemirbulatov committed:
Add method to schedule copy instructions in BlockScheduling.schedule(), removed
ReschedMap from BoUpSLP::scheduleBlock().

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -816,6 +854,9 @@ class InstructionsState {
   Instruction *MainOp = nullptr;
   Instruction *AltOp = nullptr;
 
+  /// True if alterative operation is copy instruction.
+  bool IsAltOpCopy = false;
+
```
I don't think you need this flag. Instead, InstructionsState should be initialized for the homogeneous instructions, not heterogeneous. And build the instruction, using the correctly built operands for the node

dtemirbulatov replied:
Not sure how to implement without introducing a new state somewhere? The goal was to avoid creating a new scalar instruction which we don't need in the end. Maybe we can create a new instruction and then delete it if we decide not to proceed in the end? probably we could simplify the change even more.

alexey-bataev replied:
You don't need new instructions. InstructionsState should not have any changes, but it should return Opcode/MainOp/AltOp/AltOpcode the same for all instructions/values in the list. Instead of modifying it better to implement a preliminary analysis  in buildTree_rec function, which will create homogeneous InstructionsState and build correct list of Operands, which later will be translated into new tree nodes

dtemirbulatov replied:
ok, I thought of this too. I just wanted to make sure before choosing a way. Thanks Alexey.

dtemirbulatov replied:
Done.

dtemirbulatov committed:
Resolved comments by removing IsAltOpCopy flag from InstructionsState. Resotored
original logic to handle just homogeneous operations in getSameOpcode(). Removed
checkCopyableInnerDep() with replacing functionality() with replacing with
schedular.

dtemirbulatov committed:
Fix formatting.

dtemirbulatov committed:
Resolved comments.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1088,6 +1144,69 @@ static InstructionsState getSameOpcode(ArrayRef<Value *> VL,
+  unsigned Opcode = MainOp->getOpcode();
+  unsigned AltOpcode = Opcode;
+  if (MainOp && VectorizeCopyable && all_of(VL, IsaPred<Instruction>)) {
+    for (Value *V : VL) {
+      Instruction *I = cast<Instruction>(V);
+      if (I->isIntDivRem() || I->isFPDivRem())
+        return InstructionsState::invalid();
+      if (isa<PHINode>(I)) {
+        AltOp = nullptr;
+        break;
+      }
+      unsigned VOpcode = I->getOpcode();
+      if (VOpcode != Opcode) {
+        if (AltOpcode == Opcode) {
+          AltOpcode = VOpcode;
+          AltOp = I;
+        }
+        if (VOpcode != AltOpcode) {
+          AltOp = nullptr;
+          break;
+        }
+      }
+    }
```
Better to extract into a lambda/function

dtemirbulatov replied:
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1125,6 +1244,25 @@ static bool doesInTreeUserNeedToExtract(Value *Scalar, Instruction *UserInst,
 
+// Determine that the vector could be vectorized with copyable elements.
+static bool isCopyableOp(ArrayRef<Value *> VL, Value *Main, Value *Alt) {
+  if (any_of(VL, IsaPred<PoisonValue>) || Main == Alt ||
```
Better to move `any_of(VL, IsaPred<PoisonValue>)` n the end of condition

dtemirbulatov replied:
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1125,6 +1244,25 @@ static bool doesInTreeUserNeedToExtract(Value *Scalar, Instruction *UserInst,
+// Determine that the vector could be vectorized with copyable elements.
+static bool isCopyableOp(ArrayRef<Value *> VL, Value *Main, Value *Alt) {
+  if (any_of(VL, IsaPred<PoisonValue>) || Main == Alt ||
+      !isa<BinaryOperator>(Main) || !isa<Instruction>(Alt) ||
```
Also, why `Alt` should be an instruction? I thought we should support non-instructions here too.

dtemirbulatov replied:
Not sure what could be done here. We are not interested in a mono-kind of operations or mono-kind operation with a non instruction value?

dtemirbulatov committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

dtemirbulatov committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

dtemirbulatov commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -594,6 +598,40 @@ static std::optional<unsigned> getElementIndex(const Value *Inst,
+       I->getOpcode() == Instruction::AShr ||
+       I->getOpcode() == Instruction::BitCast ||
+       I->getOpcode() == Instruction::Call ||
+       // Issue with scheduling with isVectorLikeInstWithConstOps
+       // operations.
```
The schedular says : extractelements/extractvalues don't need to " "be scheduled

alexey-bataev replied:
Then drop them completely from the list

dtemirbulatov commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1473,6 +1611,7 @@ class BoUpSLP {
     ScalarToTreeEntries.clear();
     MustGather.clear();
     NonScheduledFirst.clear();
+    CopyableAltOp.clear();
```
Done.

dtemirbulatov commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -2488,8 +2627,16 @@ class BoUpSLP {
-                                 APO, false};
+          Instruction *Inst = cast<Instruction>(VL[Lane]);
+          if (Inst->getOpcode() != MainOp->getOpcode() &&
+              OpIdx > (Inst->getNumOperands() - 1)) {
```
Done.

dtemirbulatov commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -3416,7 +3566,7 @@ class BoUpSLP {
 
     /// Some of the instructions in the list have alternate opcodes.
-    bool isAltShuffle() const { return S.isAltShuffle(); }
+    bool isAltShuffle() const { return S.isAltShuffle() && !IsAltOpCopy; }
```
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1088,6 +1144,69 @@ static InstructionsState getSameOpcode(ArrayRef<Value *> VL,
+      Instruction *I = cast<Instruction>(V);
+      if (I->isIntDivRem() || I->isFPDivRem())
+        return false;
+      if (isa<PHINode>(I)) {
+        return false;
+      }
```
1. Drop extra parens.
2. Why PHI cannot be treated as copyable?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1125,6 +1244,24 @@ static bool doesInTreeUserNeedToExtract(Value *Scalar, Instruction *UserInst,
   }
 }
 
+// Determine that the vector could be vectorized with copyable elements.
```
```suggestion
/// Determine that the vector could be vectorized with copyable elements.
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -1125,6 +1244,24 @@ static bool doesInTreeUserNeedToExtract(Value *Scalar, Instruction *UserInst,
+      any_of(VL, IsaPred<PoisonValue, PHINode>))
+    return false;
+
+  Instruction *MainOp = cast<Instruction>(Main);
+  Instruction *AltOp = cast<Instruction>(Alt);
```
```suggestion
  auto *MainOp = cast<Instruction>(Main);
  auto *AltOp = cast<Instruction>(Alt);
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -4158,30 +4362,32 @@ class BoUpSLP {
         // reordered during buildTree(). We therefore need to get its operands
         // through the TreeEntry.
-        if (TreeEntry *TE = BundleMember->TE) {
+        if (TreeEntry *TE = BundleMember->TE; TE && !TE->isAltOpCopy()) {
```
Not sure this is correct. You still need to check the deps for the operands, even if the node is marked as copyable. Non-copyable instructions should go through this code.

dtemirbulatov replied:
Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -7687,6 +7930,59 @@ static bool isAlternateInstruction(const Instruction *I,
+      MainAltOps.push_back(I);
+    }
+  }
+  if (AltOps.size() > 0 && MainAltOps.size() > 0)
+    return true;
+
+  return false;
```
```suggestion
  return !AltOps.empty() && !MainAltOps.empty();
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -7697,6 +7993,8 @@ bool BoUpSLP::areAltOperandsProfitable(const InstructionsState &S,
                            Opcode0, Opcode1, OpcodeMask))
     return true;
   SmallVector<ValueList> Operands;
+  if (S.getMainOp()->getNumOperands() != S.getAltOp()->getNumOperands())
```
Why?

dtemirbulatov replied:
There was illigal operand number before, I think I resolved now. Done.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8412,6 +8708,19 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
+      }
+    }
+  } else if (VectorizeCopyable && canRepresentAsCopyable(S, VL))
+    IsAltOpCopy = true;
```
Should be in braces

dtemirbulatov replied:
Done.

dtemirbulatov committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

dtemirbulatov committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

dtemirbulatov committed:
Update llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

dtemirbulatov committed:
Resolved comments.

dtemirbulatov commented:
ping.

alexey-bataev commented:
It must be rebased. Plus, I'm currently working on scheduling reworking, which, most probably, won't require scheduling model update in this patch, and it can be dropped.
Also, I mentioned that current approach misses some corner cases, like non-instructions, phis, etc. It should be addressed
